### PR TITLE
Introduce i.MX RT FlexSPI HyperRAM driver

### PR DESCRIPTION
**Pull request by pdgendt**
Add the FlexSPI HyperBUS driver to support HyperRAM external devices.

Technically not a flash driver so it currently doesn't have the flash API implemented. Can be added, if required.

Tested with [Cypress S27KS0641](https://www.cypress.com/file/183506/download) and following `dts` config:
```
&flexspi {
 	status = "okay";
	reg = <0x402a8000 0x4000>, <0x60000000 DT_SIZE_M(8)>;
	combination-mode;
	sck-differential-clock;
	rx-clock-source = <3>; // External input from DQS pad

	s27ks0641: s27ks0641@0 {
		status = "okay";
		compatible = "nxp,imx-flexspi-hyperram";
		reg = <0>;
		size = <67108864>;
		label = "S27KS0641";
		spi-max-frequency = <332000000>;

		cs-interval-unit = <1>;
		cs-interval = <2>;
		cs-setup-time = <4>;
		cs-hold-time = <0>;
		data-valid-time = <1>;
		column-space = <3>;
		word-addressable;
		ahb-write-wait-unit = <2>;
		ahb-write-wait-interval = <0>;
	};
};
```

### CODE REVIEW

**Comment by pdgendt** [comment]
[This application note](https://www.nxp.com/docs/en/nxp/application-notes/AN12239.pdf) specifies how to improve read/write performance with MPU settings and `ahb` buffer/prefetch features.
But I don't know how to do the MPU setting properly and enabling `ahb-prefetch` on my custom board resulted in bus faults so there is probably something missing here to fully utilize the HyperRAM.

**Comment by carlescufi** [comment]
@sylvioalves copying you here since you mentioned the same issue to me recently

**Comment by carlescufi** [comment]
@erwango does STM32 have an equivalent? i.e. RAM connected over SPI, but then directly addressable after.

**Comment by erwango** [comment]
> @erwango does STM32 have an equivalent? i.e. RAM connected over SPI, but then directly addressable after.

Yes, we also have OCTOSPI IPs supporting HyperRAM memories in memory mapped mode.

**Comment by carlescufi** [comment]
API meeting 2021.03.30:

- Move `flash_mcux_flexspi.c` and `flash_mcux_flexspi_hyperram.c` to `drivers/memc`
- Leave `flash_mcux_flexspi_nor.c` in `drivers/flash`

File: drivers/memc/Kconfig.mcux [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2020 NXP
+# Copyright (c) 2021 Basalte bv
+# SPDX-License-Identifier: Apache-2.0
+
+if HAS_MCUX_FLEXSPI
+
+config MEMC_MCUX_FLEXSPI_HYPERRAM
+	bool "MCUX FlexSPI HyperRAM driver"
+	select MEMC_MCUX_FLEXSPI
```
Comment (MaureenHelm) [inline drivers/memc/Kconfig.mcux:9]: nit: these lines should be added in commit `drivers: memc: Introduce i.MX RT FlexSPI HyperRAM driver`

File: drivers/memc/memc_mcux_flexspi.c [hunk 2]
```diff
@@ -6,44 +6,44 @@
...
-struct flash_flexspi_config {
+struct memc_flexspi_config {
 	FLEXSPI_Type *base;
 	uint8_t *ahb_base;
 	bool ahb_bufferable;
 	bool ahb_cacheable;
 	bool ahb_prefetch;
 	bool ahb_read_addr_opt;
 	bool combination_mode;
+	bool sck_differential_clock;
```
Comment (MaureenHelm) [inline drivers/memc/memc_mcux_flexspi.c:23]: nit: this should be added in commit `drivers: memc: Introduce i.MX RT FlexSPI HyperRAM driver`

File: west.yml [hunk 3]
```diff
@@ -100,7 +100,7 @@ manifest:
       revision: 41132e9220f8bc1223084975350c5e5f3b492afe
       path: tools/net-tools
     - name: hal_nxp
-      revision: b916bca1d5976b0157be80b326e3bb46ef605286
+      revision: pull/83/head
```
Comment (MaureenHelm) [inline west.yml:103]: This can be updated now that the hal PR has been merged

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thanks @pdgendt !

File: drivers/memc/Kconfig.mcux (hunk 1 above)
Comment (pdgendt) [reply drivers/memc/Kconfig.mcux:9]: Thanks for the review, this got lost in the rebase :)

**Review by MaureenHelm — APPROVED** [review]

File: include/drivers/memc/memc_mcux_flexspi.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * Copyright 2020 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <sys/types.h>
+#include <fsl_flexspi.h>
+
+int memc_flexspi_update_lut(const struct device *dev, uint32_t index,
```
Comment (galak) [inline include/drivers/memc/memc_mcux_flexspi.h:11]: Are these APIs meant for application level use?

**Review by galak — CHANGES_REQUESTED** [review]
Overall looks good, only issue is with ` include/drivers/memc/memc_mcux_flexspi.h`

I think this should be in `drivers/memc/memc_mcux_flexspi.h` and we should add:

`zephyr_include_directories(${ZEPHYR_BASE}/drivers)` to `soc/arm/nxp_imx/rt/CMakeLists.txt`

File: include/drivers/memc/memc_mcux_flexspi.h (hunk 4 above)
Comment (MaureenHelm) [reply include/drivers/memc/memc_mcux_flexspi.h:11]: No

Comment (pdgendt) [reply include/drivers/memc/memc_mcux_flexspi.h:11]: Wouldn't the following make more sense in `drivers/flash/CMakeLists.txt`?
```
zephyr_include_directories_ifdef(
  CONFIG_FLASH_MCUX_FLEXSPI_NOR
  ${ZEPHYR_BASE}/drivers/memc
)
```

**Comment by pdgendt** [comment]
> Overall looks good, only issue is with ` include/drivers/memc/memc_mcux_flexspi.h`
> 
> I think this should be in `drivers/memc/memc_mcux_flexspi.h` and we should add:
> 
> `zephyr_include_directories(${ZEPHYR_BASE}/drivers)` to `soc/arm/nxp_imx/rt/CMakeLists.txt`

Wouldn't the following make more sense in `drivers/flash/CMakeLists.txt`?
```
zephyr_include_directories_ifdef(
  CONFIG_FLASH_MCUX_FLEXSPI_NOR
  ${ZEPHYR_BASE}/drivers/memc
)
```

**Comment by galak** [comment]
> > Overall looks good, only issue is with ` include/drivers/memc/memc_mcux_flexspi.h`
> > I think this should be in `drivers/memc/memc_mcux_flexspi.h` and we should add:
> > `zephyr_include_directories(${ZEPHYR_BASE}/drivers)` to `soc/arm/nxp_imx/rt/CMakeLists.txt`
> 
> Wouldn't the following make more sense in `drivers/flash/CMakeLists.txt`?
> 
> ```
> zephyr_include_directories_ifdef(
>   CONFIG_FLASH_MCUX_FLEXSPI_NOR
>   ${ZEPHYR_BASE}/drivers/memc
> )
> ```

It might, I guess the question is who is calling those APIs?

**Comment by pdgendt** [comment]
> It might, I guess the question is who is calling those APIs?

I think only flash or memc drivers will. @MaureenHelm, can you confirm?

**Comment by MaureenHelm** [comment]
> > It might, I guess the question is who is calling those APIs?
> 
> I think only flash or memc drivers will. @MaureenHelm, can you confirm?

That's correct

**Review by galak — APPROVED** [review]

**Pull request merged**